test/show.jl

```diff
@@ -818,3 +811,20 @@ end
     @test sprint(show, Val(:Float64)) == "Val{:Float64}()" # Val of a symbol
     @test sprint(show, Val(true))     == "Val{true}()"     # Val of a value
 end
+
+@testset "printing of Pair's" begin
+    for (p, s) in (Pair(1.0,2.0)                          => "1.0 => 2.0",
+                   Pair(Pair(1,2), Pair(3,4))             => "(1=>2) => (3=>4)",
+                   Pair{Integer,Int64}(1, 2)              => "Pair{Integer,Int64}(1, 2)",
+                   (Pair{Integer,Int64}(1, 2) => 3)       => "(1=>2) => 3",
```

Comment:
I think I remember why I had made this choice: 
```
julia> [1 => 2, 1=>'2']
2-element Array{Pair{Int64,Any},1}:
 Pair{Int64,Any}(1, 2)
 Pair{Int64,Any}(1, '2')
```
The type specification looks redundant. Maybe `compact` should use the arrow (loosing info), but nested pairs should not be printed compactly by default, i.e. 
``` 
julia> 1=>2=>3
1 => (2 => 3)

julia> Pair{Any,Any}(1, 2) => 3
Pair{Any,Any}(1, 2) => 3
```
?
(PS: with #22981, the initial redundant example would still show non-compactly though).